### PR TITLE
AzureFleet public preview changes

### PR DESCRIPTION
**Pull request by SahilArora92**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

spec_pr_review_workflow_diagram

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [x] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

### CODE REVIEW

**Comment by SahilArora92** [comment]
Private preview PR is ARM signed off: https://github.com/Azure/azure-rest-api-specs-pr/pull/14241

**Comment by SahilArora92** [comment]
https://github.com/Azure/azure-rest-api-specs-pr/pull/14241
https://github.com/Azure/azure-rest-api-specs-pr/pull/18079

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+using TypeSpec.Http;
+using TypeSpec.Rest;
+using TypeSpec.Versioning;
+using TypeSpec.OpenAPI;
+using Azure.ResourceManager;
+
+namespace Microsoft.AzureFleet;
+
+@doc("An AzureFleet's virtualMachineScaleSet")
+model VirtualMachineScaleSet {
```
Comment (mentat9) [inline specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:16]: >model VirtualMachineScaleSet [](http://example.com/codeflow?start=0&length=28)

Use VirtualMachineScaleSet is ProxyResource<VirtualMachineScaleSetProperties> with a separate VirtualMachineScaleSetProperties definition that declares just the RP-specific properties. This automates the common ARM declarations.

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  @visibility("read")
+  @path
+  name: string;
+
+  @doc("""
+  The compute RP resource id of the virtualMachineScaleSet 
+  "subscriptions/{subId}/resourceGroups/{rgName}/providers/Microsoft.Compute/virtualMachineScaleSets/{vmssName}"
+  """)
+  @visibility("read")
+  id: string;
```
Comment (mentat9) [inline specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:29]: >id [](http://example.com/codeflow?start=2&length=2)

The name, id and type properties will be defined for you by typespec if you use the ProxyResource<> template.

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+  @removed(Versions.v2024_05_01_preview)
+  @doc("This represents the Provisioning State of the VMSS in response to the last operation that was performed on it by Azure Fleet resource.")
+  @visibility("read")
+  provisioningState: ProvisioningState;
```
Comment (mentat9) [inline specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:34]: >provisioningState [](http://example.com/codeflow?start=2&length=17)

These three properties (provisioningState, operationStatus, and error) will be correctly declared in a "properties" object if you use the ProxyResource<template>

File: specification/azurefleet/AzureFleet.Management/fleet.tsp [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+model ManagedServiceIdentityUpdate
+  is OptionalProperties<UpdateableProperties<Azure.ResourceManager.Foundations.ManagedServiceIdentity>>;
+
+model ResourcePlanUpdate
+  is OptionalProperties<UpdateableProperties<Azure.ResourceManager.Foundations.Plan>>;
+
+@doc("The status of the current operation.")
+@Azure.Core.lroStatus
+union ProvisioningState {
```
Comment (mentat9) [inline specification/azurefleet/AzureFleet.Management/fleet.tsp:56]: >ProvisioningState [](http://example.com/codeflow?start=6&length=17)

I think you can include the built-in ResourceProvisioningState (for the standard/required states) and just add your additional states here. Just something I noticed, not required or anything.

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp (hunk 1 above)
Comment (SahilArora92) [reply specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:16]: I tried making it a proxy resource but we don't need the extra fields which come with it. It is just a custom List operation.
image
Even Compute has the same for custom List Operation: https://learn.microsoft.com/en-us/rest/api/compute/virtual-machine-scale-sets/list-skus?view=rest-compute-2024-03-01&tabs=HTTP

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp (hunk 2 above)
Comment (SahilArora92) [reply specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:29]: See my other comment.

File: specification/azurefleet/AzureFleet.Management/fleet.tsp (hunk 4 above)
Comment (SahilArora92) [reply specification/azurefleet/AzureFleet.Management/fleet.tsp:56]: Updated.

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  operationStatus: ProvisioningState;
+
+  @added(Versions.v2024_05_01_preview)
+  @doc("Error Information when `operationStatus` is `Failed`")
+  @visibility("read")
+  error?: ApiError;
+}
+
+@doc("The response of a VirtualMachineScaleSet list operation.")
+model VirtualMachineScaleSetListResult {
```
[the remark this answers is not in the document]
Comment (SahilArora92) [reply specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:48]: We want a custom List function.

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp (hunk 1 above)
Comment (SahilArora92) [reply specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:16]: Also if we did define VirtualMachineScaleSet as a proxy resource then we would have to implement get for a single VirtualMachineScaleSet as well. We don't want to follow that pattern.

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp (hunk 3 above)
Comment (SahilArora92) [reply specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:34]: See my other comment https://github.com/Azure/azure-rest-api-specs/pull/29038#discussion_r1605493293

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp (hunk 1 above)
Comment (SahilArora92) [reply specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:16]: @mentat9 Please have a look.

Comment (mentat9) [reply specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:16]: Per offline agreement, the full ProxyResource model will be added in the next API version. To support the the team's investment in this existing API version, ARM will sign off with just the list API, updated to match the ARM ProxyResource LIST format. The next API version will add the missing point GET API.

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp (hunk 2 above)
Comment (mentat9) [reply specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:29]: Resolved per offline discussion.

File: specification/azurefleet/AzureFleet.Management/examples/2023-11-01-preview/VirtualMachineScaleSets_ListByFleet_MaximumSet.json [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    "subscriptionId": "7B0CD4DB-3381-4013-9B31-FB6E6FD0FF1C",
+    "resourceGroupName": "rgazurefleet",
+    "name": "myFleet"
+  },
+  "responses": {
+    "200": {
+      "body": {
+        "value": [
+          {
+            "id": "/subscriptions/7B0CD4DB-3381-4013-9B31-FB6E6FD0FF1C/resourceGroups/rgazurefleet/providers/Microsoft.Compute/virtualMachineScaleSets/myVmss",
```
Comment (mentat9) [inline specification/azurefleet/AzureFleet.Management/examples/2023-11-01-preview/VirtualMachineScaleSets_ListByFleet_MaximumSet.json:15]: >Compute [](http://example.com/codeflow?start=119&length=7)

AzureFleet

File: specification/azurefleet/AzureFleet.Management/examples/2024-05-01-preview/VirtualMachineScaleSets_ListByFleet.json [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    "subscriptionId": "7B0CD4DB-3381-4013-9B31-FB6E6FD0FF1C",
+    "resourceGroupName": "rgazurefleet",
+    "name": "myFleet"
+  },
+  "responses": {
+    "200": {
+      "body": {
+        "value": [
+          {
+            "id": "/subscriptions/7B0CD4DB-3381-4013-9B31-FB6E6FD0FF1C/resourceGroups/rgazurefleet/providers/Microsoft.Compute/virtualMachineScaleSets/myVmss",
```
Comment (mentat9) [inline specification/azurefleet/AzureFleet.Management/examples/2024-05-01-preview/VirtualMachineScaleSets_ListByFleet.json:15]: >Compute [](http://example.com/codeflow?start=119&length=7)

AzureFleet

File: specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp (hunk 5 above)
[the remark this answers is not in the document]
Comment (mentat9) [reply specification/azurefleet/AzureFleet.Management/virtualMachineScaleSet.tsp:52]: Resolved per offline discussion.

**Comment by mentat9** [comment]
@SahilArora92 - Signing off for ARM with comments. Please check my comments on a couple of the examples.

File: specification/azurefleet/AzureFleet.Management/examples/2024-05-01-preview/VirtualMachineScaleSets_ListByFleet.json (hunk 7 above)
Comment (SahilArora92) [reply specification/azurefleet/AzureFleet.Management/examples/2024-05-01-preview/VirtualMachineScaleSets_ListByFleet.json:15]: Thanks.

**Review by rkmanda — APPROVED** [review]

**Pull request merged**